feat: add pointers concept (#823)
exercism/cpp: concepts/pointers/introduction.md

> # Introduction

Author: ```suggestion
# Introduction
```

About.md is missing. Please fix before merge.

Like many other languages, C++ has pointers.

> You already know `references` and pointers are similar, but think of them as a level closer to the inner workings of your computer.

Reviewer: Here and in the rest of the PR: the use of backticks looks inconsistent. I'd suggest using backticks for code and italics or quotes for technical terms that you want to highlight.

Pointers are variables that hold memory addresses.
They are used to directly interact with memory, enabling dynamic memory allocation and efficient data manipulation in C++.

If you're new to pointers, they can feel a little mysterious but once you get used to them, they're quite straight-forward.

They're a crucial part of C++, so take some time to really understand them.
The bare-bone version in this concept is also called _dumb pointer_ or _raw pointer_.
With modern C++ there are also `smart pointers`, the basic type is not smart at all and you have to handle all the work manually.

Before digging into the details, it's worth understanding the use of pointers.
Pointers are a way to share memory with other parts of our program, which is useful for two major reasons:
1. When you have large amounts of data, making copies to pass between functions is very inefficient.
  By passing the memory location of where the data is stored instead, you can dramatically reduce the resource-footprint of our programs.

> 2. By passing pointers between functions, you can access and modify the single copy of the data directly, meaning that any changes made by one function are immediately visible to other parts of the program when the function ends.

Reviewer: ```suggestion
2. By passing pointers between functions, you can access and modify the single copy of the data directly, meaning that any changes made by one function are immediately visible to other parts of the program.
```

not only then, think threads

> 

Reviewer: Points 1 and 2 can also be achieved through references. You could add two more bullet points that explain that pointers can be reassigned to point to a different object, and they can be a null-pointer, i.e. not pointing to any object.

## General Syntax

> A pointer declaration in C++ involves specifying the data type the pointer points to, followed by an asterisk (`*`) and the pointer's name.

Reviewer: ```suggestion
A pointer declaration in C++ involves specifying the data type to which the the pointer is pointing, followed by an asterisk (`*`) and the pointer's name.
```

sounds a little bit better to my (German) ear

When pointers are declared, they are not automatically initialized.
Without explicit assignment, a pointer typically holds an indeterminate value, often referred to as a "garbage address."
While certain compilers might initialize pointers to `nullptr`, this behavior is not guaranteed across all compilers, so it's essential not to rely on it.
It's best practice to explicitly initialize raw pointers and verify their non-null status before utilization to avoid potential issues.

```cpp

> int* ptr = nullptr; // Declares a pointer and makes sure it is not invalid

Reviewer: In the other concepts you've used brace-initialization instead of this assignment-like syntax.

```

To assign the address of a variable to a pointer, you use the address-of operator (`&`).

> Dereferencing a pointer is done using the asterisk (`*`) operator.

Reviewer: The standard call it the "unary `*` operator". They explain that it performs an "indirection" so if you're looking for a name you could call it the "indirection operator".

```cpp
std::string opponent{"Solomon Lane"};
// 'ethan' points to the address of the string opponent
std::string* ethan{&opponent}; 
// Instead of ethan's, the opponent's name address is given to the passPort
std::string passportName{*ethan};
```

Attention: dereferencing has to be done explicitly, while _references_ just worked like an alias.

## Pointer Arithmetic

_Pointer arithmetic_ allows you to perform arithmetic operations on pointers, which is particularly useful when working with arrays.
Adding an integer to a pointer makes it point to a different element.

```cpp

> // Stargate addresses

Reviewer: Since this concept is about pointers and addresses using the term "stargate addresses" might be a little bit confusing (one is the technical term related to the concept, the other is the fictional term for interstellar travel.)

int gateAddresses[] = {462, 753, 218, 611, 977};
// 'ptr' points to the first element of 'gateAddresses'

> int* ptr{gateAddresses}; 

Reviewer: To understand this line one has to know about how arrays decay to pointers.  
I'd suggest `&gateAdresses[0]` instead of `gateAddresses`.  
That doesn't change anything else but I think it's easier to understand.

> Adding an integer to a pointer moves it to a different memory location based on the size of the data type it points to.

Reviewer: I think this sentence can be hard to understand for beginners. How about something like "makes it point to a different element"?

// Accesses the third Stargate address through pointer arithmetic
int dialedAddress{*(ptr + 2)}; 
// Chevron encoded! Dialing Stargate address:
openStarGate(dialedAddress);
```

~~~~exercism/caution
Pointer arithmetic in C++ can easily lead to __undefined behavior__ if not handled carefully.
Undefined behavior can manifest in unexpected program outcomes, crashes, or even security vulnerabilities.
One infamous example of the consequences of undefined behavior occurred in the [explosion of the Ariane 5 rocket][ariane-flight-v88] in 1996, where a software exception caused by the conversion of a 64-bit floating-point number to a 16-bit signed integer led to a catastrophic failure.
~~~~

## Dynamic Memory Allocation and Null Pointers

In C++, `new` is used to dynamically allocate memory for objects on the heap during runtime.
It returns a pointer to the allocated memory address, allowing objects to have lifetimes not tied to a specific scope.
However, it's important to always pair `new` with `delete` to properly deallocate memory.

> Dynamic memory allocation should be used cautiously to prevent common issues like memory leaks and accessing deallocated memory.

Reviewer: I would even use stronger language to discourage the use of "naked" or "raw" `new`.

Here's a brief example, with a `struct` `Superhero` that has a member variable `superpower`.
The main function creates a pointer `dianaPrince` to a `Superhero` object (representing Wonder Woman).
The `->` operator is used to access the member variable `superpower`, showcasing Wonder Woman's iconic "Lasso of Truth."

```cpp
struct Superhero {
    std::string superpower;
};

> Superhero* dianaPrince = new Superhero;

Reviewer: Do you really want to use `new` before it's introduced in the next section?

// Reset pointer to null, symbolizing there is no more darkMarkBearer at the moment
darkMarkBearer = nullptr; 
```

A pointer should either hold the address of a valid object, or be set to `nullptr`.
That way you only need to test pointers for null, and can assume any non-null pointer is valid.
When an object is destroyed, any pointers to the destroyed object will be left _dangling_ (they will not be automatically set to `nullptr`).
It is your responsibility to detect these cases and ensure those pointers are subsequently set to `nullptr`.

~~~~exercism/advanced
In older code, you might encounter two alternatives to `nullptr`.
Firstly, the literal `0` is specifically interpreted as a null value for pointers, though it's the only scenario where an integral literal can be assigned to a pointer.

> Secondly, the `preprocessor macro` `NULL`, inherited from C and defined in the `<cstddef>` header, is another representation of a null pointer, though its usage is less common in modern C++ code.

Reviewer: Is the term "preprocessor macro" in backticks on purpose?
